src/main/drivers/flash_w25n01g.c

```diff
@@ -0,0 +1,756 @@
+    fdevice->geometry.totalSize = fdevice->geometry.sectorSize * fdevice->geometry.sectors;
+
+    fdevice->couldBeBusy = true; // Just for luck we'll assume the chip could be busy even though it isn't specced to be
+
+    w25n01g_deviceReset(fdevice->busdev);
+
+    // Upper 4MB (32 blocks * 128KB/block) will be used for bad block replacement area.
+
+    // Blocks in this area are only written through bad block LUT,
+    // and factory written bad block marker in unused blocks are retained.
+
+    // When a replacement block is required,
+    // (1) "Read BB LUT" command is used to obtain the last block mapped,
+    // (2) blocks after the last block is scanned for a good block,
+    // (3) the first good block is used for replacement, and the BB LUT is updated.
+
+    // There are only 20 BB LUT entries, and there are 32 replacement blocks.
+    // There will be a least chance of running out of replacement blocks.
+    // If it ever run out, the device becomes unusable.
+
+#if 0
+    // Protection to upper 1/32 (BP[3:0] = 0101, TB=0), WP-E on
+    //w25n01g_writeRegister(fdevice->busdev, W25N01G_PROT_REG, (5 << 3)|(0 << 2)|(1 << 1));
+
+    // No protection, WP-E on
+    w25n01g_writeRegister(fdevice->busdev, W25N01G_PROT_REG, (0 << 3)|(0 << 2)|(1 << 1));
+
+    // Continuous mode (BUF = 0), ECC enabled (ECC = 1)
+    w25n01g_writeRegister(fdevice->busdev, W25N01G_CONF_REG, W25N01G_CONFIG_ECC_ENABLE);
+#endif
+
+#if 0
+    // XXX Should be gone in production
+    uint8_t sr1, sr2, sr3;
+    sr1 = w25n01g_readRegister(fdevice->busdev, W25N01G_PROT_REG);
+    sr2 = w25n01g_readRegister(fdevice->busdev, W25N01G_CONF_REG);
+    sr3 = w25n01g_readRegister(fdevice->busdev, W25N01G_STAT_REG);
+
+    debug[1] = sr1;
+    debug[2] = sr2;
+    debug[3] = sr3;
+
+    DPRINTF(("Detect: PROT 0x%x CONF 0x%x STAT 0x%x\r\n", sr1 & 0xff, sr2 & 0xff, sr3 & 0xff));
+#endif
+
+    w25n01g_deviceInit(fdevice);
+
+    fdevice->vTable = &w25n01g_vTable;
+
+    return true;
+}
+
+/**
+ * Erase a sector full of bytes to all 1's at the given byte offset in the flash chip.
+ */
+void w25n01g_eraseSector(flashDevice_t *fdevice, uint32_t address)
+{
+    const uint8_t cmd[] = { W25N01G_INSTRUCTION_BLOCK_ERASE, 0, W25N01G_LINEAR_TO_PAGE(address) >> 8, W25N01G_LINEAR_TO_PAGE(address) & 0xff };
+
+    w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_BLOCK_ERASE_MS);
+
+    w25n01g_writeEnable(fdevice);
+
+    ENABLE(fdevice->busdev);
+    spiTransfer(fdevice->busdev->busdev_u.spi.instance, cmd, NULL, sizeof(cmd));
+    DISABLE(fdevice->busdev);
+}
+
+//
+// W25N01G does not support full chip erase.
+// Call eraseSector repeatedly.
+
+void w25n01g_eraseCompletely(flashDevice_t *fdevice)
+{
+    for (uint32_t block = 0; block < fdevice->geometry.sectors; block++) {
+        w25n01g_waitForReady(fdevice, W25N01G_TIMEOUT_BLOCK_ERASE_MS);
+
+        // Issue erase block command
+        w25n01g_writeEnable(fdevice);
```

Comment:
Are the calls to `w25n01g_waitForReady` and `w25n01g_writeEnable` call needed here as `w25n01g_eraseSector` also calls `w25n01g_waitForReady` and `w25n01g_writeEnable`?